### PR TITLE
Tp04

### PR DESCRIPTION
**Pull request by mihaicostin34**
### Pull Request Overview

This pull request adds TP4

### Testing Strategy

This pull request was tested by running `npm run start`

### TODO or Help Wanted

This pull request still needs an opinion on the theory part.

### Build

- [x] Ran `npm build`.

### Author

Signed-off-by: Mihai Costin <mihaicostin342@gmail.com>

### CODE REVIEW

File: docs/tp/04/index.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+- 01 + 0011
+- \+ 01
+- \+
+
+:::warning
+  Même si vous voyez l’opérateur arithmétique '+', sachez qu’ici il ne s’agit pas de l’addition! On parle strictement des chaînes de caractères, donc on a besoin exactement du caractère '+'.
+:::
+
+## Grammaires avec ANTLR4
+À partir de ce TP, on va utiliser ANTLR pour l’analyse du texte. ANTLR (Another Tool For Language Recognition) est un outil pour generer des analyseurs (parsers) qui permettent le traitement, l’exécution et la traduction du code. On va employer ANTLR pour la partie de front-end de notre compilateur.
```
Comment (Ioan-Cristian) [inline docs/tp/04/index.md:48]:
```suggestion
À partir de ce TP, on va utiliser ANTLR pour l’analyse du texte. ANTLR (Another Tool For Language Recognition) est un outil pour générer des analyseurs (parsers) qui permettent le traitement, l’exécution et la traduction du code. On va employer ANTLR pour la partie front-end de notre compilateur.
```

File: docs/tp/04/index.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+Pour les grammaires indépendantes du contexte, ANTLR utilise des fichiers `.g4`. Pour écrire une grammaire, il faut créer un fichier avec le même nom que la grammaire que vous souhaitez définir. 
+
+### Jétons
+En ANTLR, on peut même utiliser des productions pour décrire les jétons. Dans ce cas, le corps de la production doit contenir que des jétons. On va identifier les jétons avec des noms qui commencent avec des lettres **MAJUSCULES** (On vous conseile d’utiliser que des majuscules pour les jétons, pour pouvoir les distinguer plus facilement). D’habitude, on trouve les définitions des jétons à la fin de la grammaire.
```
Comment (Ioan-Cristian) [inline docs/tp/04/index.md:53]:
```suggestion
En ANTLR, on peut utiliser des productions pour décrire des jétons. Dans ce cas, le corps de la production doit contenir que des jétons. On va identifier les jétons avec des noms qui commencent avec des lettres **MAJUSCULES** (On vous conseile d’utiliser que des majuscules pour les jétons, pour pouvoir les distinguer plus facilement). D’habitude, on trouve les définitions des jétons à la fin de la grammaire.
```

File: docs/tp/04/index.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+p : s '+'s;
+s : ZERO s ONE  
+  | ZERO ONE;
+ZERO: '0';
+ONE: '1';  
+```
+
+## Exercices
+1. Ouvrez les fichiers du laboratoire. Dans TP04/ex1.g4, completez les `TODO` pour écrire une grammaire qui reconnaît une instruction de déclaration d'une variable, ayant la syntaxe suivante:
```
Comment (Ioan-Cristian) [inline docs/tp/04/index.md:99]:
```suggestion
1. Ouvrez les fichiers du laboratoire. Dans TP04/ex1.g4, complétez les `TODO` pour écrire une grammaire qui reconnaît une instruction de déclaration d'une variable, ayant la syntaxe suivante:
```

File: docs/tp/04/index.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+```c
+  <nom_du_type> <nom_de_la_variable> ;
+
+  //Exemples:
+  int a;
+  char b23;
+  long mnop1;
+```
+Où:
+- le nom du type est l'un des options suivantes: `int`, `long`, `bool`, `char`
```
Comment (Ioan-Cristian) [inline docs/tp/04/index.md:109]:
```suggestion
- le nom du type est l'une des options suivantes: `int`, `long`, `bool`, `char`
```

File: docs/tp/04/index.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+- le nom de la variable est une chaîne qui peut contenir des lettres et des nombres, mais ne peut pas commencer avec un nombre
+
+:::warning
+  Faites attention aux `;`. Ils doivent toujours apparaître dans la déclaration!
+:::
+
+2. Ouvrez les fichiers du laboratoire. Dans TP04/ex2.g4, completez les `TODO` pour écrire une grammaire qui reconnaît des expréssions mathématiques avec des nombres entières et des noms de variables.
```
Comment (Ioan-Cristian) [inline docs/tp/04/index.md:116]:
```suggestion
2. Ouvrez les fichiers du laboratoire. Dans TP04/ex2.g4, complétez les `TODO` pour écrire une grammaire qui reconnaît des expréssions mathématiques avec des entiers et des noms de variables.
```

**Comment by Ioan-Cristian** [comment]
Peut-être changer le nom de la demande d'ajout (pull request) en TP04 ? :)

File: docs/tp/04/index.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
+---
+sidebar_position: 5
+description: Définir les langages de programmation
+slug: tp/04
+---
+
+# 04- Grammaires indépendantes de contexte
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:7]:
```suggestion
# 04 - Grammaires indépendantes de contexte
```

File: docs/tp/04/index.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  - le symbole de production
+  - une chaîne de zéro ou plusieurs terminaux et variables (appelée **le corps de la production**) qui représente un moyen de décrire la variable située dans la tête de la production. Cette chaîne peut aussi être une **expression régulière** construite en utilisant les variables et les terminaux.
+
+### Exemple
+Pour definir un langage qui accepte les expréssions des sommes entre deux nombres contenant des 0 consécutives suivis par des 1 consécutives:
+$$
+  P \rightarrow S + S \\
+  S \rightarrow 0S1 \\
+  S \rightarrow \epsilon 
+$$
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:30]:
```suggestion
$$
  P \rightarrow S + S
$$

$$
  S \rightarrow 0S1
$$

$$
  S \rightarrow \epsilon 
$$
```

File: docs/tp/04/index.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+Dans cet exemple, il y a:
+- les terminaux: 0 et 1
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:32]:
```suggestion
- les terminaux: $$\{+, 0, 1\}$$
```

File: docs/tp/04/index.md [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+- les non-terminaux: P, S
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:33]:
```suggestion
- les non-terminaux: $$\{P, S\}$$
```

File: docs/tp/04/index.md [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+- le symbole de début: P
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:34]:
```suggestion
- le symbole de début: $$P$$ 
```

File: docs/tp/04/index.md [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+- les terminaux: 0 et 1
+- les non-terminaux: P, S
+- le symbole de début: P 
+- trois productions
+
+Chaînes acceptées:
+- 00001111 + 0011
+- 01 + 0011
+- \+ 01
+- \+
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:41]:
```suggestion
- `00001111+0011`
- `01+0011`
- `+01`
- `+`
```

File: docs/tp/04/index.md [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+### Jétons
+En ANTLR, on peut même utiliser des productions pour décrire les jétons. Dans ce cas, le corps de la production doit contenir que des jétons. On va identifier les jétons avec des noms qui commencent avec des lettres **MAJUSCULES** (On vous conseile d’utiliser que des majuscules pour les jétons, pour pouvoir les distinguer plus facilement). D’habitude, on trouve les définitions des jétons à la fin de la grammaire.
+
+```antlr4
+  INTEGER: (-)?[0-9]+ ;  //un jéton pour représenter les nombres entières
+  WHITESPACE: ' ' ; //un jéton pour représenter les espaces blancs
+  VARIABLE_NAME: [a-zA-Z]+ //un jéton pour représenter les noms des variables
+```
+
+### Variables
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:61]:
```suggestion
### Règles
```

File: docs/tp/04/index.md [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+
+On va identifier les variables avec des lettres **minuscules**.
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:63]:
```suggestion
On va identifier les règles avec des lettres **minuscules**.
```

File: docs/tp/04/index.md [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+:::warning
+En ANTLR, on **ne peut pas avoir des productions avec la même tête**. Ce qu’on peut faire, c’est de mettre les deux corps dans un seul et de les séparer avec l’opérateur `|`.
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:66]:
```suggestion
En ANTLR, on **ne peut pas avoir des règles avec la même tête**. Ce qu’on peut faire, c’est de mettre les deux corps dans un seul et de les séparer avec l’opérateur `|`.
```

File: docs/tp/04/index.md [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+:::
+
+```antlr4
+  assignment: VARIABLE_NAME '=' INTEGER ; //une variable pour représenter une affectation
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:70]:
```suggestion
  assignment: VARIABLE_NAME '=' INTEGER ; //une règle pour représenter une affectation
```

File: docs/tp/04/index.md [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+  
+  //une variable pour représenter une expression d'addition
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:72]:
```suggestion
  //une règle pour représenter une expression d'addition
```

File: docs/tp/04/index.md [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+          | INTEGER
+          ;
+```
+
+### Grammaires
+Prenons la grammaire précedante et essayons de l’écrire avec ANTLR.
+
+:::warning
+La grammaire sera un peu différente, à cause du fait qu’en ANTLR on ne peut pas utiliser la chaîne vide. Alors, pour ne pas nous eloigner de l’exemple, S va commencer
+par la valeur `01`.
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:83]: You can have the empty rule:

```antlr4
grammar demo;

ZERO: '0';
ONE: '1'; 

p : s '+' s;
s : ZERO s ONE  
  | ; 
```

I might be wrong about this, please try it out before changing.

File: docs/tp/04/index.md [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+La grammaire devient: 
+
+```antlr4
+grammar demo;
+
+p : s '+'s;
+s : ZERO s ONE  
+  | ZERO ONE;
+ZERO: '0';
+ONE: '1';
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:95]: I would writing the tokens first.

File: docs/tp/04/index.md (hunk 3 above)
Comment (alexandruradovici) [inline docs/tp/04/index.md:99]:
```suggestion
1. Ouvrez les fichiers du laboratoire. Dans `TP04/ex1.g4`, completez les `TODO` pour écrire une grammaire qui reconnaît une instruction de déclaration d'une variable, ayant la syntaxe suivante:
```

File: docs/tp/04/index.md (hunk 5 above)
Comment (alexandruradovici) [inline docs/tp/04/index.md:116]:
```suggestion
2. Ouvrez les fichiers du laboratoire. Dans `TP04/ex2.g4`, completez les `TODO` pour écrire une grammaire qui reconnaît des expréssions mathématiques avec des nombres entières et des noms de variables.
```

File: docs/tp/04/index.md [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+- le nom du type est l'un des options suivantes: `int`, `long`, `bool`, `char`
+- le nom de la variable est une chaîne qui peut contenir des lettres et des nombres, mais ne peut pas commencer avec un nombre
+
+:::warning
+  Faites attention aux `;`. Ils doivent toujours apparaître dans la déclaration!
+:::
+
+2. Ouvrez les fichiers du laboratoire. Dans TP04/ex2.g4, completez les `TODO` pour écrire une grammaire qui reconnaît des expréssions mathématiques avec des nombres entières et des noms de variables.
+
+3. Ouvrez les fichiers du laboratoire. Dans TP04/ex3.g4, combinez les deux grammaires des exercices précédents pour obtenir une grammaire qui accepte les instructions suivantes:
```
Comment (alexandruradovici) [inline docs/tp/04/index.md:118]:
```suggestion
3. Ouvrez les fichiers du laboratoire. Dans `TP04/ex3.g4`, combinez les deux grammaires des exercices précédents pour obtenir une grammaire qui accepte les instructions suivantes:
```

File: docusaurus.config.ts [hunk 20]
```diff
@@ -140,9 +140,9 @@ const config: Config = {
...
       theme: prismThemes.github,
       darkTheme: prismThemes.dracula,
-      additionalLanguages: ['nasm'],
+      additionalLanguages: ['nasm', `antlr4`],
     },
   } satisfies Preset.ThemeConfig,
 };
 
-export default config;
+export default config;
```
Comment (alexandruradovici) [inline docusaurus.config.ts:148]:
```suggestion
export default config;

```

Add newline.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Review by alexandruradovici — APPROVED** [review]

**Pull request merged**